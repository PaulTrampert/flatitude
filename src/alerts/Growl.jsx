import React from 'react';
import PropTypes from 'prop-types';
import ColorTypes from '../colors/ColorTypes.js';
import classnames from 'classnames';

class Growl extends React.Component {
  render() {
    let {
      children,
      type
    } = this.props;
    return (
      <div className={classnames("growl", type)}>
        <div className="icon">
        </div>
        <div className="message">
          {children}
        </div>
        <div className="dismiss">
          <button>
          </button>
        </div>
      </div>
    );
  }
}

Growl.propTypes = {
  children: PropTypes.node,
  type: ColorTypes.isRequired
};

export default Growl;