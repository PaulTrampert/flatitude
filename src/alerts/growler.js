import EventBroker from "../util/EventBroker.js";

class Growler {
  constructor(events) {
    this.events = events || new EventBroker('growler');
    this.nextId = 0;
  }

  growl = ({type, message, onDismiss}) => {
    this.events.publish({
      id: this.nextId,
      type,
      message,
      onDismiss: onDismiss || (() => {})
    });
    this.nextId++;
  }

  subscribe = (handler) => {
    return this.events.subscribe(handler);
  }
}

const instance = new Growler();

export {
  instance as default,
  Growler
};