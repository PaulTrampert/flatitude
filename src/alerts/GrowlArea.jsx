import React from 'react';
import growler from './growler.js';
import Growl from './Growl.jsx';
import RenderInBody from '../util/RenderInBody.jsx';

class GrowlArea extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      growls: []
    };
  }

  componentDidMount = () => {
    this.unsub = growler.subscribe(this.showGrowl);
  }

  componentWillUnmount = () => {
    this.unsub && this.unsub();
    delete this.unsub;
  }

  showGrowl = (growl) => {
    let growls = this.state.growls.slice();
    growls.push(growl);
    this.setState({
      growls
    });
  }

  handleDismiss = (id) => {
    let growls = this.state.growls.slice();
    let idx = growls.findIndex(g => g.id === id);
    let growl = growls.splice(idx, 1)[0];
    growl.onDismiss();
    this.setState({
      growls
    });
  }

  render() {
    let {
      growls
    } = this.state;

    return (
      <RenderInBody>
        <div className="growl-area">
          {
            growls.map(g => (
              <Growl key={g.id} id={g.id} type={g.type} onDismiss={this.handleDismiss}>
                {g.message}
              </Growl>
            ))
          }
        </div>
      </RenderInBody>
    );
  }
}

export default GrowlArea;