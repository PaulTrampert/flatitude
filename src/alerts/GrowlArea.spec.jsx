import React from 'react';
import {shallow} from 'enzyme';
import GrowlArea from './GrowlArea.jsx';
import Growl from './Growl.jsx';
import growler from './growler.js';

jest.mock('./growler.js');

describe('GrowlArea', () => {
  let subject;
  let unsubscribe;
  let growls;
  let instance;

  beforeEach(() => {
    unsubscribe = jest.fn();
    growler.subscribe.mockReturnValue(unsubscribe);
    growls = [
      {
        id: 1,
        type: 'info',
        message: 'herp',
        onDismiss: jest.fn()
      },
      {
        id: 2,
        type: 'danger',
        message: 'derp',
        onDismiss: jest.fn()
      },
    ];

    subject = shallow(<GrowlArea />, {disableLifecycleMethods: true});
    instance = subject.instance();
  });

  describe('constructor', () => {
    it('initializes state', () => {
      expect(subject.state()).toEqual({
        growls: []
      });
    });
  });

  describe('componentDidMount', () => {
    it('subscribes to the growler', () => {
      instance.componentDidMount();

      expect(growler.subscribe).toHaveBeenCalledWith(instance.showGrowl);
    });

    it('keeps the unsubscribe function', () => {
      instance.componentDidMount();

      expect(instance.unsub).toBe(unsubscribe);
    });
  });

  describe('componentWillUnmount', () => {
    beforeEach(() => {
      instance.componentDidMount();
    });

    it('unsubscribes from the growler', () => {
      instance.componentWillUnmount();
      expect(unsubscribe).toHaveBeenCalled();
    });

    it('deletes the unsub function', () => {
      instance.componentWillUnmount();
      expect(instance.unsub).not.toBeDefined();
    });
  });

  describe('showGrowl', () => {
    let growl;
    beforeEach(() => {
      growl = {
        id: 1,
        type: 'info',
        message: 'flerp',
        onDismiss: jest.fn()
      };

      subject.setState({growls});
    });

    it('adds the growl to the list of growls', () => {
      instance.showGrowl(growl);

      expect(subject.state().growls).toEqual([
        ...growls,
        growl
      ]);
    });
  });

  describe('handleExpire', () => {
    beforeEach(() => {
      subject.setState({growls});
    });

    it('expires the correct growl', () => {
      instance.handleExpire(2);

      expect(subject.state().growls).toEqual([
        growls[0]
      ]);
    });
  });

  describe('handleDismiss', () => {
    beforeEach(() => {
      subject.setState({growls});
    });

    it('dismisses the correct growl', () => {
      instance.handleDismiss(2);

      expect(subject.state().growls).toEqual([
        growls[0]
      ]);
    });

    it("calls the growl's onDismiss function", () => {
      instance.handleDismiss(2);

      expect(growls[1].onDismiss).toHaveBeenCalled();
    });
  });

  describe('render', () => {
    describe('when there are no growls', () => {
      it('renders an empty growl area in the body', () => {
        expect(subject).toMatchSnapshot();
      });
    });

    describe('when there are growls', () => {
      beforeEach(() => {
        subject.setState({growls});
      });

      it('renders the growls in the growl area', () => {
        expect(subject).toMatchSnapshot();
      });

      it('binds onDismiss for each growl', () => {
        expect(subject.find(Growl).at(0).props().onDismiss).toBe(instance.handleDismiss);
        expect(subject.find(Growl).at(1).props().onDismiss).toBe(instance.handleDismiss);
      });

      it('binds onExpire for each growl', () => {
        expect(subject.find(Growl).at(0).props().onExpire).toBe(instance.handleExpire);
        expect(subject.find(Growl).at(1).props().onExpire).toBe(instance.handleExpire);
      });
    });
  });
});
