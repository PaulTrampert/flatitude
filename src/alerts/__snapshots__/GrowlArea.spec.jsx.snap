// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrowlArea render when there are growls renders the growls in the growl area 1`] = `
<RenderInBody>
  <div
    className="growl-area"
  >
    <Growl
      durationMs={10000}
      id={1}
      key="1"
      onDismiss={[Function]}
      onExpire={[Function]}
      type="info"
    >
      herp
    </Growl>
    <Growl
      durationMs={10000}
      id={2}
      key="2"
      onDismiss={[Function]}
      onExpire={[Function]}
      type="danger"
    >
      derp
    </Growl>
  </div>
</RenderInBody>
`;

exports[`GrowlArea render when there are no growls renders an empty growl area in the body 1`] = `
<RenderInBody>
  <div
    className="growl-area"
  />
</RenderInBody>
`;
